flashlight/fl/nn/modules/Container.h

```diff
@@ -40,16 +40,66 @@ class FL_API Container : public Module {
 
   Container();
 
+  /**
+   * Removes all modules and parameters from the container.
+   */
+  virtual void clear();
+
+  /**
+   * Find orphaned params (i.e. params not in modules contained in the modules_
+   * list). This can be used to preserve the order of orphaned params when
+   * copying/cloning a container. std::unordered_multimap<module_idx, param_idx>
+   * The module_idx is used to identify after which module params should be
+   * serted and the param_idx is used to index the specific param. The following
+   * example demonstrates its usage by ensuring params and modules are inserted
+   * in the same order when making a copy:
+   * \code
+      void copy(const MyContainer& other) {
+        auto orphanParamIdxMap = other.getOrphanedParamsIdxMap();
+        for (int i = -1; i < static_cast<int>(other.modules_.size()); ++i) {
+          if (i >= 0) {
+            add(other.modules_[i]->clone());
+          }
+          auto [paramIter, pEnd] = orphanParamIdxMap.equal_range(i);
+          for (; paramIter != pEnd; ++paramIter) {
+            const auto& param = other.params_[paramIter->second];
+            params_.emplace_back(param.copy());
+          }
+        }
+      }
+     \endcode
+   *
+   * A module_idx of -1 indicates the orphaned params are to be inserted
+   * before the first module
+   *
+   * @return A multimap of orphaned params and the module index they appear
+   * after
+   */
+  std::unordered_multimap<int, int> getOrphanedParamsIdxMap() const;
+
  public:
   /**
-   * Adds a module to a `Container` by making a copy of the underlying module.
-   * Note that parameters are still shared, due to Variable's copy semantics.
+   * Adds a module to a `Container` by making a copy of the underlying module if
+   * an lvalue or moving it if and rvalue
+   *
+   * @param[in] module the module to add.
+   */
+  template <typename T>
+  void add(T&& module) {
+    static_assert(
+        !std::is_lvalue_reference_v<T>,
+        "add() can only accept rvalues. Use std::move().");
+    add(std::make_shared<std::decay_t<T>>(std::move(module)));
```

Comment:
I believe as the static assert checks for an lvalue reference, `std::forward<T>` isn't strictly required and `std::move` is sufficient - or rather in this case they function the same. I think I used `std::move` instead as it was slightly less to type. I'll change to avoid confusion.